### ISSUE TITLE
Numpy access to ancestral and derived state columns in TreeSequence

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/tskit#2632]
It would be really nice if we could access the ancestral and derived states as numpy string arrays. In the common single-allele case, we can easily do this:
```python
ancestral_state = tables.sites.ancestral_state.view("S1")
Out[64]: array([b'T', b'G', b'A', ..., b'T', b'T', b'A'], dtype='|S1')
```
However, this is a byte array and "not recommended". I guess we should try to return the values as proper numpy ``str`` types?

Can easily do this but it makes a copy (I assume?)
```python
derived_state.astype(str)
Out[63]:
array(['T', 'G', 'A', ..., 'T', 'T', 'A'], dtype='<U1')
```

I haven't managed to make a direct view of the memory that gets treated as a unicode string.

As a first pass we could just error out of the states aren't all single chars (ascii chars even), and I guess we could just copy the data in a way that's numpy string friendly when we have variable length alleles? I guess it would be good to know if this is possible before painting ourselves into a corner.

Related: #2631

**Comment by benjeffery** [comment]
` tables.sites.ancestral_state.view("S1").astype(str, copy=False)` works for me, and I think avoids the copy?

**Comment by jeromekelleher** [comment]
No, this doesn't work unfortunately - I think ``copy=False`` is a request that numpy does its best to fulfil, when possible. I just tried it out there and the OWNDATA flag is false when you do ``.astype(str, copy=False)``.

I don't get why the view ``dtype=U1`` doesn't work though - aren't these UTF-8 encoded single byte strings?

**Comment by benjeffery** [comment]
> I don't get why the view `dtype=U1` doesn't work though - aren't these UTF-8 encoded single byte strings?

`numpy.dtype("U1").itemsize` is `4` as there has to be room for the largest, 4-byte unicode character.

`numpy.dtype("S1").itemsize` is `1`. So I don't think we can get to a unicode dtype without a copy or some low-level hacking.

**Comment by jeromekelleher** [comment]
Note that the current plan for this is to make a numpy stringdype in C
